packages/mongodb/src/index.ts

```diff
@@ -0,0 +1,148 @@
+/* eslint-disable @typescript-eslint/no-non-null-assertion */
+import type {
+  Adapter,
+  AdapterSession,
+  AdapterUser,
+  VerificationToken,
+} from "next-auth/adapters"
+import type * as MongoDB from "mongodb"
+import { Account } from "next-auth"
+
+export const collections = {
+  Users: "users",
+  Accounts: "accounts",
+  Sessions: "sessions",
+  VerificationTokens: "verification_tokens",
+}
+
+export const format = {
+  /** Takes a mongoDB object and returns a plain old JavaScript object */
+  from<T = Record<string, unknown>>(object: Record<string, any>): T {
+    const newObject: Record<string, unknown> = {}
+    for (const key in object) {
+      const value = object[key]
+      if (key === "_id") {
+        newObject.id = value.toHexString()
+      } else {
+        newObject[key] = value
+      }
+    }
+    return newObject as T
+  },
+}
+
+export function MongoDBAdapter(options: {
+  db: MongoDB.Db
+  ObjectId: typeof MongoDB.ObjectId
+}): Adapter {
+  const { db: m, ObjectId } = options
+  const { from } = format
+
+  // Converts from string to ObjectId
+  const _id = (hex?: string) => {
+    if (hex?.length !== 24) return new ObjectId()
+    return new ObjectId(hex)
+  }
+
+  const { Users, Accounts, Sessions, VerificationTokens } = {
+    Users: m.collection<AdapterUser>(collections.Users),
+    Accounts: m.collection<Account>(collections.Accounts),
+    Sessions: m.collection<AdapterSession>(collections.Sessions),
+    VerificationTokens: m.collection<VerificationToken>(
+      collections.VerificationTokens
+    ),
+  }
+  return {
+    async createUser(data) {
+      const user = { _id: _id(), ...(data as any) }
+      await Users.insertOne(user)
+      return from<AdapterUser>(user)
+    },
+    async getUser(id) {
+      const user = await Users.findOne({ _id: _id(id) })
+      if (!user) return null
+      return from<AdapterUser>(user)
+    },
+    async getUserByEmail(email) {
+      const user = await Users.findOne({ email })
+      if (!user) return null
+      return from<AdapterUser>(user)
+    },
+    async getUserByAccount(provider_providerAccountId) {
+      const account = await Accounts.findOne(provider_providerAccountId)
+      if (!account) return null
+      const user = await Users.findOne({ _id: _id(account.userId) })
+      if (!user) return null
+      return from<AdapterUser>(user)
+    },
+    async updateUser(data) {
+      const { value: user } = await Users.findOneAndUpdate(
+        { _id: _id(data.id) },
+        { $set: data }
+      )
+      return from<AdapterUser>(user!)
+    },
+    async deleteUser(id) {
+      await Promise.all([
+        m.collection(collections.Accounts).deleteMany({ userId: id }),
```

Comment:
The relations doesn't have the string converter. Does that mean that the relations are string primitives?

Reply:
Could you explain what you mean here exactly? 😊

Reply:
The **ObjectId** consists of a 12-byte value (4-byte is a creation timestamp, 5-byte is a random value, 3-byte is a initially random incrementing counter), but its hexadecimal representation (the string), depending on the charset you're using, can consist of around 1-byte per character while having 24 characters which means double the size.

The **ObjectId** not only have a better performance for indexing, but also takes less space if you're not indexing. For me converting is a must because of those two reasons.

Reply:
Fixed!

Reply:
There's one thing I forgot to mention that might be a problem, I'm not really sure how are you handling this on other cases.

The **ObjectId** is not serializable by NextJS. So if you try to pass it on the `getServerSideProps`, `getInitialProps`, and `getStaticProps` it will throw an error.

Since I already use `mongodb` on my application, I handle that by transforming my data before dehydrating it.

On the TypeORM providers for example, do you return a lean version of the objects or do you return a TypeORM model? If you just return the TypeORM model, it might be a good idea to transform the **ObjectId** to hexadecimal before returning or add a note on the documentation.

---

Some references:

https://github.com/blitz-js/superjson/issues/69
https://github.com/blitz-js/superjson#using-with-nextjs
https://github.com/blitz-js/superjson/pull/71

Disclaimer: I'm not suggesting to use `superjson` or anything, it's just the closest thing I found to help explain the issue.

Reply:
Oh well, it looks you're doing that already. Nice!

Reply:
Yes, before sending it to the core, I transform the ids to hex strings.